Add internal definitions for S3 ls recursive

shaunrd0 opened the pull request:
This adds internal definitions for `sm::VFS::ls_recursive` and `S3::ls_filtered` for S3 only. Results are collected using a `S3Scanner` class that manages fetching results from S3, and advances `LsScanIterator` until the next accepted result is reached. The `S3Scanner::iterator()` method returns an iterator for the scan which can be passed to STL constructors or algorithms using input iterators.

---
TYPE: FEATURE
DESC: Add internal definitions for S3 ls recursive

shaunrd0 commented on tiledb/sm/filesystem/vfs.h:
```diff
@@ -161,11 +170,17 @@ struct VFSBase {
+
+  inline S3* s3() {
+    return &s3_;
+  }
```
@eric-hughes-tiledb just wanted to point this out for review, I added this to simplify S3 unit tests a bit in my last commit.

I think this could go away once file systems inherit from a common base class. When that is done I could write a test class that stores a pointer to the FilesystemBase and construct what I need for the test based on the input URI. That said if we don't want to add this accessor for now, I could instead construct `sm::S3` as a member of `S3Test` class and compile it out with preprocessor checks for non-s3 builds until base class work is done.

eric-hughes-tiledb reviewed (changes requested):
For as much feedback as I've left, I consider this PR in pretty good shape.
* `class VFSTest` needs to be made more compliant with C.41. It can't entirely yet because it needs to deal with "not supported" compile environments
* There's some other smallish polishing items.

The most significant thing is going to be to change the callback type to a template argument. This is done all over `<algorithm>`, for example. This will ease the next step of hooking this code up by wrapping the C callback that will come in through the API.

shaunrd0 commented on test/support/src/vfs_helpers.cc:
```diff
@@ -415,5 +442,132 @@ std::string TemporaryDirectoryFixture::create_temporary_array(
+          &io_,
+          &compute_,
+          tiledb::test::create_test_config())
+    , temp_dir_(tiledb::test::test_dir(prefix)) {
```
I didn't have a use for `ensure_supported()` at the moment since the only IO functions are part of test setup which goes back to my other comment on not being able to construct the test classes without failing TEMPLATE_LIST_TEST_CASE from throwing in functions called by the constructors. I can still add it if we want to leave it unused and document it's purpose.

The is_supported member and rename is done though, and I added a note to throw in the `VFSTestBase` constructor when we can produce a list of supported filesystems and pass to Catch2.

shaunrd0 commented on test/support/src/vfs_helpers.cc:
```diff
@@ -415,5 +442,132 @@ std::string TemporaryDirectoryFixture::create_temporary_array(
+    [[maybe_unused]] const sm::URI& uri,
+    [[maybe_unused]] size_t count,
+    [[maybe_unused]] const std::string& prefix) {
+#ifdef HAVE_S3
```
This specific function was removed in my last commit, but I wasn't able to get constexpr working in other similar functions for S3Test class. When I made the switch to constexpr I hit errors from the undefined `VFS::s3` accessor with no S3 symbol to define it with when S3 is not enabled, but it sounds like that shouldn't really matter.

Based on these docs it seems the discarded statements are checked in non-template functions, but glad to take another pass if you have an idea to get this working.

> Outside a template, a discarded statement is fully checked. if constexpr is not a substitute for the #if preprocessing directive
https://en.cppreference.com/w/cpp/language/if

shaunrd0 commented on test/support/src/vfs_helpers.cc:
```diff
@@ -415,5 +442,132 @@ std::string TemporaryDirectoryFixture::create_temporary_array(
+    vfs_.s3()->flush_object(object_uri).ok();
+    expected_results_.emplace_back(object_uri.to_string(), data.size());
+  }
+#endif
```
If I move `setup_test()` into the constructor and throw here I wouldn't be able to construct `S3Test` without throwing for builds without S3 enabled, so the TEMPLATE_LIST_TEST_CASE would fail before we leave the constructor. I removed `create_objects` since it's only called by `setup_test` and used preproccessor checks to turn that into a stub if the backend is not enabled by the build. I left a separate comment on why I used preproccessor instead of constexpr.

eric-hughes-tiledb replied:
>I wouldn't be able to construct S3Test without throwing for builds without S3 enabled

Naively, yes, that would be the case. We've already solved exactly this problem with `struct S3_within_VFS`, from which `class VFS` gets its `s3_` member variable.. We'll need to use exactly this solution for `class VFSTest`. I think we should be able to use the same `struct `S3_within_VFS`.

shaunrd0 commented on tiledb/sm/filesystem/ls_callback.h:
```diff
@@ -0,0 +1,53 @@
+ * @return `1` if the walk should continue to the next object, `0` if the walk
+ *    should stop, and `-1` on error.
+ */
+using LsCallback = std::function<int32_t(const char*, size_t, uint64_t, void*)>;
```
I ended up using the C++ interface for `LsCallbackWrapper::operator()` class from another comment here. I added a C++ typedef but it's currently not in use, the operator() has this signature.

shaunrd0 commented on tiledb/sm/filesystem/ls_callback.h:
```diff
@@ -0,0 +1,53 @@
+ * @param object_size The size of the object at the path.
+ * @param data Cast to user defined struct to store paths and offsets.
+ * @return `1` if the walk should continue to the next object, `0` if the walk
+ *    should stop, and `-1` on error.
```
Added this behavior in `LsCallbackWrapper::operator()`.

eric-hughes-tiledb replied:
OK, but the interface definition here still needs to change. We need to be throwing on error. This is the interface description for a C callback function. It's not the description of anything that's a C++ function or function object.

shaunrd0 commented on tiledb/sm/filesystem/vfs.h:
```diff
@@ -427,6 +442,17 @@ class VFS : private VFSBase, S3_within_VFS {
+   * @param cb The callback to invoke on each object collected.
+   * @param data Data passed to the callback used to store collected results.
+   */
+  void ls_recursive(const URI& parent, LsCallback cb, void* data) const;
```
I took a pass at this today and tests are passing but was a little unsure if what I did here is what you had in mind, I expect we'll need to move some things around a bit.

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_callback.h:
```diff
@@ -0,0 +1,87 @@
+/**
+ * Wrapper for the C API ls callback function and it's associated data.
+ */
+class LsCallbackWrapper {
```
The name of this class should reflect that it's only for use with the C API

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_callback.h:
```diff
@@ -0,0 +1,87 @@
+#include <functional>
+#include <stdexcept>
+
+template <class F>
```
This declaration needs a comment that it's a stub and stating in informal language what the concept is supposed to represent, which is that requires `operator()` with a particular signature. We can flesh out the concept in another PR to avoid delaying this one further.

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_callback.h:
```diff
@@ -0,0 +1,87 @@
+/**
+ * @file ls_callback.h
```
This file needs a new name, and some of its contents do also.

`callback` is a word specific to C. We don't want to be using it for C++ objects. We will need a constructor argument called `callback` to handle the C API (and a related member variable), but that's about the extent to which we want to use this word.

With C++ `<algorithm>` they're not called "callbacks" because they usually do _not_ call back into the caller, even as they call into other code provided by the call. Look, say, at https://en.cppreference.com/w/cpp/algorithm/copy in the signature of `copy_if`, where the template argument is `UnaryPredicate`. The template arguments we use should be similarly named.

ihnorton replied:
Renaming the file is fine, but
> callback is a word specific to C

"callback" is commonly used in Javascript and many other languages; it came up multiple times during a JS UI discussion just yesterday.

https://en.wikipedia.org/wiki/Callback_(computer_programming)

https://en.wikipedia.org/wiki/Callback_(computer_programming)

eric-hughes-tiledb commented on test/support/src/vfs_helpers.cc:
```diff
@@ -415,5 +438,137 @@ std::string TemporaryDirectoryFixture::create_temporary_array(
+  setup_test();
+}
+
+void VFSTest::test_ls_recursive(
```
This function should not be defined in this file at all, which likely means it doesn't belong in this class. 

Tests for the filtered lists belong in the translation units that use them. In this case it's `unit-vfs.cc`. In order to get the definition there, it means that the present `class VFSTest` should be the base class of a class derived from it and defined in `unit-vfs.cc`.

eric-hughes-tiledb commented on test/src/unit-vfs.cc:
```diff
@@ -351,3 +352,86 @@ TEST_CASE("VFS: test ls_with_sizes", "[vfs][ls-with-sizes]") {
+  DYNAMIC_SECTION(
+      fs.temp_dir_.backend_name()
+      << " ls_cb with recursion: " << (recursive ? "true" : "false")) {
+    fs.test_ls_cb(cb, recursive);
```
This statement uses a test that's defined completely elsewhere. That's generally a sign that the code is organized poorly. The code that runs this test should be in the present source file (unless it's shared by other units, which it's not in this case).

eric-hughes-tiledb commented on test/src/unit-vfs.cc:
```diff
@@ -351,3 +352,86 @@ TEST_CASE("VFS: test ls_with_sizes", "[vfs][ls-with-sizes]") {
+  }
+
+  tiledb::sm::LsCallbackCAPI cb =
+      [](const char* path, size_t path_len, uint64_t size, void* data) {
```
The selection predicates for our own tests should not participate in the C API. IOW there should be no `data` argument.

eric-hughes-tiledb commented on tiledb/sm/filesystem/s3.h:
```diff
@@ -86,6 +89,14 @@ using tiledb::common::filesystem::directory_entry;
 
 namespace tiledb::sm {
```
We need a `constexpr bool` here for S3 being enabled.

eric-hughes-tiledb commented:
Previous review isn't complete. I pressed "send" too early. To be continued...

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_callback.h:
```diff
@@ -0,0 +1,87 @@
+class LsCallbackWrapper {
+ public:
+  /** Constructor */
+  LsCallbackWrapper(LsCallbackCAPI cb, void* data)
```
The construction argument for a wrapper for a C API function should take a function pointer, not a C++ function object.

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_callback.h:
```diff
@@ -0,0 +1,87 @@
+namespace tiledb::sm {
+
+using LsCallback =
+    std::function<bool(const std::string_view&, uint64_t, void*)>;
```
Is this dead code?

shaunrd0 replied:
Yes it is, I meant to call this out in my comments but may not have. I ended up using this signature for the `LsCallbackWrapper::operator()`, but didn't use this directly. I was expecting more changes so I left it around in case we found a use for it that I wasn't seeing, but I can remove if it's still not needed.

eric-hughes-tiledb commented on test/support/src/vfs_helpers.cc:
```diff
@@ -415,5 +438,137 @@ std::string TemporaryDirectoryFixture::create_temporary_array(
+  CHECK(ls_objects == expected_results_);
+}
+
+void VFSTest::setup_test() {
```
Does anything else call this function, or can it just be the constructor body?

eric-hughes-tiledb reviewed (commented):
* I think all the `setup_test()` functions should go. If they can't become the constructor body, then we need to identify the problems with initialization and solve them properly.

eric-hughes-tiledb commented:
I really hate the way github makes it so easy to send the review early.
Still reviewing.

eric-hughes-tiledb commented on tiledb/sm/filesystem/vfs.cc:
```diff
@@ -807,6 +807,21 @@ tuple<Status, optional<std::vector<directory_entry>>> VFS::ls_with_sizes(
 }
 
+template <LsCb F>
+void VFS::ls_recursive(const URI& parent, [[maybe_unused]] F cb) const {
```
This definition should go into the header to avoid explicitly instantiation. We want to make it easy for test classes to supply entirely different functions that what we're going to use to wrap C functions for the C API.

eric-hughes-tiledb commented on tiledb/sm/filesystem/s3.cc:
```diff
@@ -921,6 +914,62 @@ tuple<Status, optional<std::vector<directory_entry>>> S3::ls_with_sizes(
 }
 
+template <LsCb F>
+void S3::ls_cb(const URI& prefix, F cb, const std::string& delimiter) const {
```
Also as below for the `class VFS` version, we should define this in the header.

eric-hughes-tiledb reviewed (changes requested):
I think we need a different interface for the function supplied as a selection predicate. The other issues are more about code structure and don't really affect this issue. Let's work on this one last.

eric-hughes-tiledb commented:
Done for now.

shaunrd0 commented:
I addressed the remaining feedback and updated unit tests today.. As a final step I ran the VCF query in the original story and the 3202 results are collected but the iteration never ends, so there's an edge case here that needs fixed and coverage added. :monocle_face:

shaunrd0 commented:
Fixed the bug I ran into before the holidays and added test coverage. I created a story for follow up to inherit from FilesystemBase since it's pure virtual and would require implementing all the methods in this PR. I also noticed some of the methods in FilesystemBase don't exist for cloud backends and vice versa, so there will likely be more refactoring for FilesystemBase once it's being used for a cloud backend and not just local Posix. I added more detail to the story with some example code to show what I'm thinking there. It's not a ton of work but seems mostly out of scope for this PR.

https://app.shortcut.com/tiledb-inc/story/38117

I'm not sure where ls_recursive fits into FilesystemBase since it's a template method that can't be virtual, so open to suggestions there. Otherwise all feedback has been addressed and this is ready for review :slightly_smiling_face:

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_scanner.h:
```diff
@@ -0,0 +1,232 @@
+ * LsScanIterator iterates over the results of ls requests wrapped by classes
+ * deriving from LsScanner. See S3Scanner as an example.
+ *
+ * @tparam T The LsScanner type that created this iterator.
```
I'd recommend renaming this template argument for clarity. For containers, short names like `T` and `U` generally refer to types of content within them. `scanner_type` would a perfectly good name.

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_scanner.h:
```diff
@@ -0,0 +1,232 @@
+   * @return The current object being visited.
+   */
+  constexpr reference operator*() const noexcept {
+    return *ptr_;
```
This will segfault if executed immediately after the default constructor. Because the value of `ptr_` is part of the representation of the end-iterator, there needs to be a check here. It should throw when the iterator is not dereferenceable.

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_scanner.h:
```diff
@@ -0,0 +1,232 @@
+ * @tparam U The data type stored by this iterator.
+ */
+template <class T, class U>
+class LsScanIterator {
```
Needs documentation about how the end iterator is represented. (It's `ptr_ == nullptr`).

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_scanner.h:
```diff
@@ -0,0 +1,232 @@
+  using reference = const U&;
+  using iterator_category = std::input_iterator_tag;
+
+  /** Default constructor. */
```
Need documentation that this constructs the end iterator.

eric-hughes-tiledb reviewed (changes requested):
* All the new tests are at the unit level. We don't want new unit tests in `tiledb_unit` (much of which is not unit tests).
  * They should appear in the test directory for the unit, which is `sm/filesystem/test`.
  * Test support code should go in the most local directory possible. In this case this is the same directory.
  * After these changes, we shouldn't need to touch the `/test` directory at all.
  * There are two `class VFS`. Probably don't need two.
* See https://en.cppreference.com/w/cpp/named_req/InputIterator.
  * The iterator doesn't need `begin()` or `end()`.
  * Need an internal function `ensure_dereferenceable()` to throw when `ptr_ == nullptr`.

eric-hughes-tiledb commented:
Damn github UI. Still reviewing.

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_scanner.h:
```diff
@@ -0,0 +1,232 @@
+   *
+   * @return Reference to this iterator after advancing to the next object.
+   */
+  LsScanIterator& operator++() {
```
`++` on an end iterator is a noop. That should be performed here rather than in `next`.

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_scanner.h:
```diff
@@ -0,0 +1,232 @@
+   *
+   * @return Copy of this iterator prior to advancing to the next object.
+   */
+  LsScanIterator operator++(int) {
```
Ditto

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_scanner.h:
```diff
@@ -0,0 +1,232 @@
+   *    Input iterators are single-pass, so we return a copy of this iterator at
+   *    it's current position.
+   */
+  LsScanIterator begin() {
```
I don't think there's a reason to implement this function, or `end()` below either. It's not meaningful for an input iterator.

eric-hughes-tiledb commented on tiledb/sm/filesystem/s3.h:
```diff
@@ -271,6 +336,199 @@ struct S3Parameters {
+    if (ptr == end_) {
+      if (more_to_fetch()) {
+        // Fetch results and reset the iterator.
+        ptr = fetch_results();
```
Note: This line is part of a defect that will only fault rarely. Don't fix it in this PR, but do document the defect in the class docs. `@section Known Defect`.

Equality on iterators is only by `ptr_` value. The value of that member variable is that of an internal AWS buffer of some sort (`... GetContents().front()`). You don't have any control over how AWS decides what to send you back. In particular, there's no way of knowing that it's a different pointer than a previous one. Therefore, it should be assumed that it might be.

To be fully correct, the pointer type needs to be compound (1) pointer to AWS buffer (as now) and also (2) batch # (or something) which is the number of times `fetch_results()` has been called. This defect only matters when two non-end pointers are compared. Fixing this for this PR would be overkill. It is extremely unlikely that this ever happens in practice. It does need to be documented, however rare.

eric-hughes-tiledb commented on tiledb/sm/filesystem/s3.h:
```diff
@@ -271,6 +336,199 @@ struct S3Parameters {
+      // If the request returned no results, we've reached the end of the scan.
+      // We hit this case when the number of objects in the bucket is a multiple
+      // of the current max_keys.
+      return nullptr;
```
Would be better to explicitly return an end-iterator here. This code assumes knowledge of how the end-iterator is constructed.

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_scanner.h:
```diff
@@ -0,0 +1,232 @@
+ * deriving from LsScanner. See S3Scanner as an example.
+ *
+ * @tparam T The LsScanner type that created this iterator.
+ * @tparam U The data type stored by this iterator.
```
Mark this comment as a point for future discussion.

It might be better to make argument `U` be an iterator itself, rather than a raw type that we're tossing around as `U*` all the time. Any change is not for this PR.

eric-hughes-tiledb commented on tiledb/sm/filesystem/s3.h:
```diff
@@ -271,6 +336,199 @@ struct S3Parameters {
+   * @param ptr Reference to the current data pointer.
+   */
+  void advance(typename Iterator::pointer& ptr) {
+    ++ptr;
```
This statement will fail when the container does not have contiguously storage. That's true for `AWS::Vector` (which is `std::vector` with an AWS allocator), but it might not be true in other situations. This doesn't need to be changed so much as documented. The documentation should be prominent, not buried somewhere in the details.

eric-hughes-tiledb commented on tiledb/sm/filesystem/s3.h:
```diff
@@ -271,6 +336,199 @@ struct S3Parameters {
+   */
+  void advance(typename Iterator::pointer& ptr) {
+    ++ptr;
+    if (ptr == end_) {
```
This way of detecting end-of-container won't work, because the address is outside the bounds of the container.

This isn't hypothetical. We've had exactly this class of defect before. It happened when an allocation was right at the end of the memory block. A computed address of this form was an illegal address and would fail by segfault in certain circumstances.

eric-hughes-tiledb commented on tiledb/sm/filesystem/s3.h:
```diff
@@ -271,6 +336,199 @@ struct S3Parameters {
+    }
+    // Update pointers to the newly fetched results.
+    begin_ = &list_objects_outcome_.GetResult().GetContents().front();
+    end_ = &list_objects_outcome_.GetResult().GetContents().back() + 1;
```
It would be more reliable to call `begin()` and `end()` here.
* `front()` has undefined behavior when the vector is empty, and while that seems unlikely here, there's no reason to tempt fate. If `begin() == end()`, we should terminate the iteration explicitly and return an end-iterator.
* Taking the address of something outside the boundaries of a vector leads to undefined behavior. Never do this. You can generate segfaults this way. 

`begin_` and `end_` are going to need to be different types.

eric-hughes-tiledb commented on tiledb/sm/filesystem/s3.h:
```diff
@@ -271,6 +336,199 @@ struct S3Parameters {
+  /** Delimiter used for ListObjects request. */
+  std::string delimiter_;
+  /** Pointers for the current objects fetched from S3. */
+  typename Iterator::pointer begin_, end_;
```
These will need to be declared a different type. It's going to mean changing the template argument to be "iterator to `U`" rather than just "`U`"

eric-hughes-tiledb commented on tiledb/sm/filesystem/s3.h:
```diff
@@ -271,6 +336,199 @@ struct S3Parameters {
+ * S3Scanner wraps the AWS ListObjectsV2 request and provides an iterator for
+ * results. If we reach the end of the current batch of results and results are
+ * truncated, we fetch the next batch of results from S3.
+ *
```
This class needs documentation about how the member variables jointly represent the iteration state. Documentation for individual member variables is insufficient for this.

eric-hughes-tiledb reviewed (changes requested):
In addition to the previous comments:
* The `end_` member variable is fundamentally broken. We can't use a pointer for the purpose that it's being put to. This is pretty specific to `begin_` and `end_`. Their details need to be changed, but that's it. They're not being used in any conceptually incorrect way.

But for code organization (previously submitted), and this new issue, everything looks pretty good. Overall code structure is definitely in a good state.

ihnorton commented:
@eric-hughes-tiledb as long as the external API is finalized here, we need to wrap this up ASAP.

eric-hughes-tiledb commented on tiledb/sm/filesystem/ls_scanner.h:
```diff
@@ -142,7 +169,9 @@ class LsScanIterator {
    */
   LsScanIterator& operator++() {
-    scanner_->next(ptr_);
+    if (++ptr_ != pointer()) {
```
This looks like a double increment, since `next()` is being called below. I'll let it pass for now since the unit tests pass, but this might be defective.

eric-hughes-tiledb reviewed (approved):
Good enough for now.

We'll need to change `class LsScanIterator` to take an iterator type instead of an underlying data type when we generalize it past the current use only in S3. It works only because the iterator type declared (from `vector`) happens to match the one in the AWS library. That won't be true in general.
